packages/server/src/lib/cache.ts

```diff
@@ -1,69 +1,22 @@
-import { logWarn } from '../utils/logging';
+import memoize from 'memoizee';
 
-type Cache = Record<string, unknown>;
-const cache: Cache = {};
+type CacheableFunc<T> = () => Promise<T>;
+type CachedFunc<T> = CacheableFunc<T> & { clear: () => void };
 
-const retryIntervalMs = 20 * 1000;
+interface CacheOptions {
+    ttlSec: number;
+    warm?: boolean;
+}
 
-/**
- * A cache that refreshes after ttlSec seconds.
- * If there are cache reads after a value has expired but before the value has been refreshed then
- * the stale data will be returned.
- *
- * If a refresh attempt fails then it will be retried after retryIntervalMs, and stale data will be returned until this
- * completes.
- *
- * @param fn        the async function for generating the data to be cached
- * @param ttlSec    time to live in seconds
- * @param key       unique cache key
- * @param warm      if true then immediately run fn
- */
 export const cacheAsync = <T>(
-    fn: () => Promise<T>,
-    ttlSec: number,
-    key: string,
-    warm = false,
-): [() => void, () => Promise<T>] => {
-    const getValue = async (): Promise<T> => {
-        if (cache[key] !== undefined) {
-            return Promise.resolve(cache[key] as T);
-        } else {
-            // First read attempt on this key. Fetch an initial value and setup the refresh scheduler
-            try {
-                const result: T = await fn();
-                cache[key] = result;
-                return Promise.resolve(result);
-            } catch (err) {
-                logWarn(`Failed to make initial request for ${key}: ${err}`);
-                return Promise.reject(
-                    new Error(`Failed to make initial request for ${key}: ${err}`),
-                );
-            } finally {
-                const scheduleRefresh = (ms: number): void => {
-                    setTimeout(async () => {
-                        try {
-                            cache[key] = await fn();
-                            scheduleRefresh(ttlSec * 1000);
-                        } catch (err) {
-                            logWarn(`Error refreshing cached value for key ${key}: ${err}`);
-                            scheduleRefresh(retryIntervalMs);
-                        }
-                    }, ms);
-                };
-
-                scheduleRefresh(ttlSec * 1000);
-            }
-        }
-    };
-
-    const reset = (): void => {
-        cache[key] = undefined;
-    };
+    fn: CacheableFunc<T>,
+    { ttlSec, warm }: CacheOptions,
+): CachedFunc<T> => {
+    const cached = memoize(fn, { promise: true, maxAge: ttlSec * 1000, preFetch: true });
```

Comment:
Do you know if `preFetch: true` means it will kick off the first fetch immediately - or is it just after expiry?

Reply:
It won't kick off the first fetch immediately. The semantics are that if the function is called 'close' to expiry (where the default value of 'close' is 33%), it will kick off a background refresh.